feature_engine/sanity_check/similar_columns.py

```diff
@@ -0,0 +1,162 @@
+
+        self.impute_with = impute_with
+        self.missing_values = missing_values
+        self.drop_if_more_columns = drop_if_more_columns
+        self.add_if_less_columns = add_if_less_columns
+        self.verbose = verbose
+
+    def _check_input(self, X: pd.DataFrame):
+        X = _is_dataframe(X)
+
+        if self.missing_values == "raise":
+            # check if dataset contains na
+            _check_contains_na(X, self.variables_)
+            _check_contains_inf(X, self.variables_)
+
+        return X
+
+    def fit(self, X: pd.DataFrame, y: pd.Series = None):
+        """Fit columns schema
+
+        Parameters
+        ----------
+
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The input dataframe
+
+        y: None
+            y is not needed for this transformer. You can pass y or None.
+
+        """
+        X = self._check_input(X)
+
+        self.variables_ = X.columns
+        return self
+
+    def _raise_errors_if_needed(
+        self,
+        _columns_to_drop: List[str],
+        _columns_to_add: List[str],
+        X: pd.DataFrame
+    ):
+
+        if not self.drop_if_more_columns and len(_columns_to_drop) > 0:
+            raise ValueError(
+                "drop_if_more_columns is set to False"
+                " and transform df contain more columns "
+                "than the df used in fit "
+                f"(found {set(X.columns)}, "
+                f"expected ({set(self.variables_)}))."
+            )
+
+        if not self.add_if_less_columns and len(_columns_to_add) > 0:
+            raise ValueError(
+                "add_if_less_columns is set to False"
+                " and transform df contain less columns "
+                "than the df used in fit "
+                f"(found {set(X.columns)}, "
+                f"expected ({set(self.variables_)}))."
+            )
+
+    def transform(self, X: pd.DataFrame, **transform_params) -> pd.DataFrame:
+        """Drops the variable that are not in the fitted dataframe and returns
+        a new dataframe with the remaining subset of variables.
+
+        If a column is in train but not in test, then the column will be created in
+        test dataset with missing value everywhere.
+
+        If a column is in test but not in train, it will be dropped.
+
+        Parameters
+        ----------
+        X: pandas dataframe
+            The input dataframe from which features will be dropped
+
+        Returns
+        -------
+        X_transformed: pandas dataframe of shape =
+             [n_samples, n_features - len(features_to_drop)]
+
+                The transformed dataframe with the same columns
+                (in same order) than the fitted dataframe
+
+        """
+        check_is_fitted(self)
+
+        X = self._check_input(X)
+
+        _columns_to_drop = list(set(X.columns) - set(self.variables_))
+        _columns_to_add = list(set(self.variables_) - set(X.columns))
```

Comment:
why the use of `set`?

Reply:
Why not, I thought it was the appropriate data structure because we want a finite, unique list of elements (and list are not unique) ?

Reply:
> 
> 
> Why not, I thought it was the appropriate data structure because we want a finite, unique list of elements (and list are not unique) ?

If we use set on X.columns, we are in essence, expecting duplicated column names  in the dataframe. Am I correct? 

This shouldn't happen, and if it does, I don't think we should have functionality to resolve it. It is up to the user to check that their dataframes make sense.

I would be happy to use set if it makes things faster, but not otherwise.